straxen/plugins/peak_processing.py

```diff
@@ -332,97 +334,283 @@ def find_n_competing(peaks, windows, fraction):
+            if len(current):
+                self.compute_shadow(current, peaks[mask_pre], split_peaks, self.exponent, array, reorder, self.getsigma(self.sigma, self.baseline, current['area']))
+            resdict[key] = array
+        res = np.zeros(len(current), self.dtype)
+        for key, names in zip(['s1', 's2', 's2_re'], 
+                              [np.dtype(self.sdtype).names, 
+                               np.dtype(self.sdtype + self.ddtype).names, 
+                               np.dtype(self.sdtype + self.ddtype).names]):
+            for name in names:
+                res[name + '_' + key] = resdict[key][name]
+        for sa in ['', 'alt_']:
+            distance = np.sqrt((res[sa + 'pre_x_s2_re'] - current['x']) ** 2 + (res[sa + 'pre_y_s2_re'] - current['y']) ** 2)
+            distance = np.where(np.isnan(distance), 2 * straxen.tpc_r, distance)
+            res[sa + 'shadow_s2_prob'] = halfcauchy.pdf(distance, scale=self.getsigma(self.sigma, self.baseline, current['area']))
+            res[sa + 'shadow_s2_corr'] = res[sa + 'shadow_s2_re'] * res[sa + 'shadow_s2_prob']
+        res['time'] = current['time']
+        res['endtime'] = strax.endtime(current)
         return res
 
+    @staticmethod
+    def getsigma(sigma, baseline, s2):
+        return sigma / np.sqrt(s2) + baseline
+
     @staticmethod
     @numba.njit
-    def compute_shadow(peaks, pre_s2_peaks, touching_windows, exponent, res):
+    def compute_shadow(peaks, pre_peaks, touching_windows, exponent, res, pos_corr, sigmas=None):
         """
-        For each peak in peaks, check if there is a shadow-casting S2 peak
-        and check if it casts the largest shadow
+        For each peak in peaks, check if there is a shadow-casting peak
+        and check if it casts the first or second largest shadow
         """
-        for p_i, p_a in enumerate(peaks):
+        for p_i, (p_a, sigma) in enumerate(zip(peaks, sigmas)):
             # reset for every peak
-            new_shadow = 0
-            s2_indices = touching_windows[p_i]
-            for s2_idx in range(s2_indices[0], s2_indices[1]):
-                s2_a = pre_s2_peaks[s2_idx]
-                if p_a['center_time'] - s2_a['center_time'] <= 0:
+            indices = touching_windows[p_i]
+            shadow_ref_bk = 0
+            for idx in range(indices[0], indices[1]):
+                s_a = pre_peaks[idx]
+                near_dt = p_a['center_time'] - s_a['center_time']
+                if near_dt <= 0:
                     continue
-                new_shadow = s2_a['area'] * (
-                        p_a['center_time'] - s2_a['center_time'])**exponent
-                if new_shadow > res['shadow'][p_i]:
+                new_shadow = s_a['area'] * near_dt**exponent
+                if pos_corr:
+                    distance = np.sqrt((p_a['x'] - s_a['x']) ** 2 + (p_a['y'] - s_a['y']) ** 2)
+                    distance = np.where(np.isnan(distance), 2 * straxen.tpc_r, distance)
```

Comment:
Why can the distance just not be defined in this case?

Reply:
There are peaks too small to have a position([codes](https://github.com/XENONnT/straxen/blob/master/straxen/plugins/position_reconstruction.py#L17)).